### ISSUE TITLE
Cancel ActionPlan

### ISSUE THREAD

**Issue by AlexisQapa** [StevenLambion/SwiftDux#12]
Hi,

What would be the best way to cancel an actionPlan ?

For example, I have a signup screen when I tap the signup button it start a networking call. Now the user is impatient and close the signup screen before the network call finishes. How can I cancel the signUpActionPlan in the view disappear callback ?

**Comment by StevenLambion** [comment]
There isn't a direct way to cancel a running action plan, but it would be nice to provide some kind of mechanism to support it. The library itself wouldn't be able to do much more than unsubscribe from the optional publisher returned from an action plan, so any internal details of the plan would need to be performed by the developer if needed.

Ideally, I could see the action dispatcher returning a cancellable subscription object when an action plan is performed.

Here's a rough solution I threw together for the interim. Instead of returning a publisher in the action plan it subscribe directly via sink. It then passes the cancellable object to the view as a @State object via a closure.

```swift
func signUpActionPlan(username: String, password: String, canceller: (AnyCancellable) -> ()) -> ActionPlan<AppState> {
  ActionPlan<AppState> { store -> () in
    canceller(webAPI.signUp(username: username, password: password).sink { newUser in
      store.send(UserAction.setCurrentUser(newUser))
    })
  }
}

struct MyView: View {

  @MappedDispatch() private var dispatch
  
  @State private var username: String = ""
  @State private var password: String = ""
  
  @State private var signUpCancellable: AnyCancellable?
  
  var body: some View {
    Group {
      /// ...signup form
      Button(action: self.signUp) { Text("Sign Up") }
    }
    .onDisappear { self.signUpCancellable?.cancel() }
  }
  
  func signUp() {
    dispatch(signUpActionPlan(username: username, password: password) {
      self.signUpCancellable = $0
    })
  }
  
}
```

**Comment by AlexisQapa** [comment]
This would solve my issue ! What do you think about having :  
view get a [Cancellable] or a "disposeBag"
dispatcher returns Cancellable?

If a cancellable is returned the action is cancellable otherwise not.
View can then, at the library level, save automatically the cancellable if not nuls.
Maybe the api would be cleaner with always returning a cancellable but with an empty cancel function.

**Comment by StevenLambion** [comment]
I like the idea of returning an AnyCancellable, which can act as an auto-disposable entity similar to using a DisposeBag by simply adding an array to the view via @State. I don't want to require developers to handle the AnyCancellable for every dispatch they perform though.

This may require a new interface to explicitly dispatch action plans that return a cancellable object separate from the existing API.

**Comment by StevenLambion** [comment]
I added a new method to ActionPlan called `sendAsCancellable(_:)` to avoid a major refactor that might present a lot of break changes.

You pass the dispatch function from @MappedDispatch to it, and it'll return a new cancellable object:

```swift
let actionPlan = signUpActionPlan(
  username: username,
  password: password
)
self.signUpCancellable = actionPlan.sendAsCancellable(dispatch)
```

It's included in v0.11.2.
